File: lib/bindings/langs/swift/Sources/BreezSDKLiquid/PluginManager.swift
import Foundation

public class PluginManager {
    fileprivate static let TAG = "PluginManager"
    private static var nwcPlugin: BindingNwcService? = nil
    fileprivate static var queue = DispatchQueue(label: "PluginManager")
    static var logger: ServiceLogger = ServiceLogger(logStream: nil)

    static func nwc(liquidSDK: BindingLiquidSdk, pluginConfigs: PluginConfigs) throws -> BindingNwcService? {
        try PluginManager.queue.sync { [] in
            if PluginManager.nwcPlugin == nil {
                if pluginConfigs.nwc == nil {
                    return nil;
                }
                logger.log(tag: TAG, line: "Starting NWC service", level: "INFO")
                PluginManager.nwcPlugin = try liquidSDK.useNwcPlugin(config: pluginConfigs.nwc!)
            }
            return PluginManager.nwcPlugin
        }
    }

    static func shutdown() {
        PluginManager.queue.sync { [] in
            if PluginManager.nwcPlugin != nil {
                logger.log(tag: TAG, line: "Shutting down NWC service", level: "INFO")
                PluginManager.nwcPlugin?.stop()
                PluginManager.nwcPlugin = nil
            }
        }
    }
}

public class PluginConfigs {
    public var nwc: NwcConfig?
    init(nwc: NwcConfig?) {
        self.nwc = nwc;
    }
}

File: lib/bindings/langs/android/lib/src/main/kotlin/breez_sdk_liquid_notification/PluginManager.kt
package breez_sdk_liquid_notification

import breez_sdk_liquid.BindingLiquidSdk
import breez_sdk_liquid.BindingNwcService
import breez_sdk_liquid.NwcConfig

class PluginManager {
  companion object {
    private const val TAG = "PluginManager"
    private var nwc: BindingNwcService? = null

    internal fun nwc(
            liquidSDK: BindingLiquidSdk,
            configs: PluginConfigs,
            logger: ServiceLogger
    ): BindingNwcService? {
      synchronized(this) {
        if (nwc == null) {
          if (configs.nwc == null) return null
          logger.log(TAG, "Starting NWC service", "DEBUG")
          nwc = liquidSDK.useNwcPlugin(configs.nwc)
          logger.log(TAG, "Successfully started NWC service", "DEBUG")
        } else {
          logger.log(TAG, "Already started NWC service", "DEBUG")
        }
        return nwc
      }
    }

    internal fun shutdown(logger: ServiceLogger) {
      synchronized(this) {
        logger.log(TAG, "Shutting down the plugin manager", "DEBUG")
        nwc?.stop()
        nwc = null
        logger.log(TAG, "Successfully shut down the plugin manager", "DEBUG")
      }
    }
  }
}

class PluginConfigs(public val nwc: NwcConfig?) {}
